### ISSUE TITLE
sym_val_lit definition too restrictive

### ISSUE THREAD

**Issue by sogaiu** [sogaiu/tree-sitter-clojure#51]
Clojure (not sure how far back), reads `###=(identity NaN)` with no errors:
```
Clojure 1.11.1
user=> (type ##NaN)
java.lang.Double
user=> ###=(identity NaN)
##NaN
user=> (type ###=(identity NaN))
java.lang.Double
```
Though Babashka doesn't seem to:
```
Babashka v1.2.175-SNAPSHOT REPL.
Use :repl/quit or :repl/exit to quit the REPL.
Clojure rocks, Bash reaches.

user=> (type ##NaN)
java.lang.Double
user=> ###=(identity NaN)
[line 1, col 19] No reader function for tag =.
```

I think this is because the current `sym_val_lit` definition is too restrictive (similar to #46 and #50): https://github.com/sogaiu/tree-sitter-clojure/blob/421546c2547c74d1d9a0d8c296c412071d37e7ca/grammar.js#L438-L441

Perhaps making that `$.sym_lit` a `$._form` would work better.

5d5ab9b25268bce95e82f19d39dcbfaff883fe17 on the `pre-0.0.12` branch is an attempt at implementing this idea.  Clojars test results remain unchanged.  Added a test at https://github.com/sogaiu/tree-sitter-clojure/commit/ddb1cfb921e79fefb56cf95d04a6f68e149063c6.

There might be other parts of the grammar that have this kind of issue.  Perhaps it's worth looking for more.

**Comment by sogaiu** [comment]
An obvious candidate to examine is `evaling_lit`: https://github.com/sogaiu/tree-sitter-clojure/blob/421546c2547c74d1d9a0d8c296c412071d37e7ca/grammar.js#L443-L450

Babashka handles `#=#=(+ 1 1)`:

```
Babashka v1.2.175-SNAPSHOT REPL.
Use :repl/quit or :repl/exit to quit the REPL.
Clojure rocks, Bash reaches.

user=> #=#=(+ 1 1)
2
```

But Clojure doesn't seem to:

```
Clojure 1.11.1
user=> #=#=(+ 1 1)
Syntax error reading source at (REPL:1:12).
Unsupported #= form
```

Although our grammar doesn't handle this currently:

```
$ echo "#=#=(+ 1 1)" | tree-sitter.ahlinc parse -
(source [0, 0] - [1, 0]
  (ERROR [0, 0] - [0, 2])
  (evaling_lit [0, 2] - [0, 11]
    value: (list_lit [0, 4] - [0, 11]
      value: (sym_lit [0, 5] - [0, 6]
        name: (sym_name [0, 5] - [0, 6]))
      value: (num_lit [0, 7] - [0, 8])
      value: (num_lit [0, 9] - [0, 10]))))
Snippet #1	0 ms	(ERROR [0, 0] - [0, 2])
```

I wonder if it makes sense to relax the current definition...

Not that the examples so far seem to be the kinds of things that are widely used / encouraged :)

**Comment by sogaiu** [comment]
Another candidate is `tagged_or_ctag_lit`: https://github.com/sogaiu/tree-sitter-clojure/blob/baf5a2b5a6a9397457650d3c00093e330efb4f7c/grammar.js#L478-L488

I thought the `$.sym_lit` might be too restrictive.

Babashka handles `# #=(identity 'uuid) "00000000-0000-0000-0000-000000000000"`:

```
Babashka v1.2.175-SNAPSHOT REPL.
Use :repl/quit or :repl/exit to quit the REPL.
Clojure rocks, Bash reaches.

user=> # #=(identity 'uuid) "00000000-0000-0000-0000-000000000000"
#uuid "00000000-0000-0000-0000-000000000000"
```

But Clojure doesn't seem to:

```
Clojure 1.11.1
user=> # #=(identity 'uuid) "00000000-0000-0000-0000-000000000000"
Syntax error reading source at (REPL:1:21).
Reader tag must be a symbol
"00000000-0000-0000-0000-000000000000"
```

Again, our grammar doesn't handle this currently:

```
$ echo "# #=(identity 'uuid) \"00000000-0000-0000-0000-000000000000\"" | tree-sitter.ahlinc parse -
(source [0, 0] - [1, 0]
  (ERROR [0, 0] - [0, 2])
  (evaling_lit [0, 2] - [0, 20]
    value: (list_lit [0, 4] - [0, 20]
      value: (sym_lit [0, 5] - [0, 13]
        name: (sym_name [0, 5] - [0, 13]))
      value: (quoting_lit [0, 14] - [0, 19]
        value: (sym_lit [0, 15] - [0, 19]
          name: (sym_name [0, 15] - [0, 19])))))
  (str_lit [0, 21] - [0, 59]))
Snippet #1	0 ms	(ERROR [0, 0] - [0, 2])
```

but may be it makes sense to relax the current definition...

**Comment by sogaiu** [comment]
Thanks for taking a look and the comments.

The examples in https://github.com/sogaiu/tree-sitter-clojure/issues/51#issuecomment-1473122490 and https://github.com/sogaiu/tree-sitter-clojure/issues/51#issuecomment-1473147688 were both just "imagined", i.e. I didn't find similar things in existing samples.

There are several considerations for these two latter rules (not the original one with the `sym_val_lit`).

1. I would also guess that Babashka's handling of:

    * [evaling_lit](https://github.com/sogaiu/tree-sitter-clojure/issues/51#issuecomment-1473122490)
    * [tagged_or_ctor_lit](https://github.com/sogaiu/tree-sitter-clojure/issues/51#issuecomment-1473147688)

    is not intentional, but we could ask @borkdude :)

2. These things all involve `#=` -- which IIUC is one of these "discouraged" constructs [1].  I originally included it because I saw it being used in some `project.clj` files, but then, as can be seen in https://github.com/sogaiu/tree-sitter-clojure/issues/50#issue-1623284525, I found an example of usage in an ordinary file.  Since it seems that "production" code makes use of the construct, I think if some real "executor" of "Clojure" code [2] processes it and it's not too much for us to support, it makes sense to support it.

3. The changes under consideration are to make the grammar "looser", which is sort-of inline with tree-sitter's "philosophy", but it also makes the grammar conceptually simpler.  If such changes resulted in more parse errors I would be more reluctant to make them, but evidence so far suggests no additional errors.

4. If we had chosen the alternate fix (of allowing `$._form` instead of just adding an additional item to `choice`) mentioned in #46, then that would have prevented #50.  So I think it's possible that the two latter changes under consideration in this issue might prevent "genuine" parse errors in the future.  A "drawback" of prevention might be we would be less likely to "detect" such usages.

Having written the above, ATM, I don't have a strong leaning one way or another.

---

[1] It may be "discouraged", but sometimes may be practicality ought to prevail...not sure in these instances whether that's the case.

[2] Babashka is its own genuine thing IMO, it may be much younger than Clojure itself, but if it ever happened that there was code that only it executed that got widespread, I think we would want to consider whether to support that sort of thing.  Having said that, I don't know of any such code ATM.

**Comment by borkdude** [comment]
Are these examples from real code or just made up? Usually I've only seen one "nesting" of `#=(expression)` and not with multiple `##`s. 

AFAIK `#=` is even undocumented and not part of the official reader documentation and theoretically might even change or disappear (although I think this is unlikely).

https://clojure.org/reference/reader#_dispatch

**Comment by borkdude** [comment]
Although it is documented in the docstring of `*read-eval*`.

**Comment by borkdude** [comment]
A nice source of information on this grep.app:

https://grep.app/search?q=%23%3D&filter[lang][0]=Clojure

I find zero results when prepending multiple `#`s.

**Comment by sogaiu** [comment]
@borkdude Thanks for stopping by :)

> Are these examples from real code or just made up? Usually I've only seen one "nesting" of #=(expression) and not with multiple ##s.

They are mde up, though inspired by a real-world usage of `#=` in the context of metadata as seen in https://github.com/sogaiu/tree-sitter-clojure/issues/50

> AFAIK #= is even undocumented and not part of the official reader documentation 

Yes.

I take it from your response that there is no intention to necessarily continue to support the current behavior in Babashka.  Is that a reasonable take?

**Comment by sogaiu** [comment]
> A nice source of information on this grep.app:

> [https://grep.app/search?q=%23%3D&filter[lang][0]=Clojure](https://grep.app/search?q=%23%3D&filter%5Blang%5D%5B0%5D=Clojure)

> I find zero results when prepending multiple #s.

Thanks for the tip.

We've been processing the large collection of "release" jars from Clojars, and didn't see any examples there.  We might also start looking through the collection identified by Dewey at some point.

**Comment by borkdude** [comment]
I double checked and this works the same in bb:

```
user=> (meta ^#=(keyword "a") [])
{:a true}
```

> I take it from your response that there is no intention to necessarily continue to support the current behavior in Babashka.

What current behavior?

**Comment by sogaiu** [comment]
> I double checked and this works the same in bb

Yes.

> What current behavior?

Babashka handles:

* `#=#=(+ 1 1)`
* `# #=(identity 'uuid) "00000000-0000-0000-0000-000000000000"`

while Clojure does not.

**Comment by borkdude** [comment]
I would say this is undefined behavior so it may or may not change, but given that this construction is fairly niche and I don't have any plans to change it, it will likely stay the same, unless there are strong indications that it does not align with the official behavior of JVM Clojure. You could double check with Alex Miller (by pinging him here, ask him on Slack or on ask.clojure).

**Comment by sogaiu** [comment]
Thanks for the clarification and suggestion.

**Comment by sogaiu** [comment]
Addressed in 5d5ab9b25268bce95e82f19d39dcbfaff883fe17.
